### DISCUSSION TITLE
How about separating a reusable base component from ant design blazor?

### DISCUSSION THREAD

**Discussion by doddgu** [ant-design-blazor/ant-design-blazor#1809] [Ideas]
How about separating a pure interactive and custom themed BlazorComponent based on Ant Design Blazor to separate style and functionality implementations ?  

The reason for creating BlazorComponent is to make it easier for the Blazor community to create more component libraries in different design languages and enrich the blazor community ecology.
So the BlazorComponent will use ```MIT License```. 

What else is there?

The BlazorComponent focus on the ```High Reusability```, ```High Stability```, ```High Performance```, ```High Plasticity```.
When we finish refactoring the BlazorComponent, we just need to add styles and specific apis to provide different implementations of design, such as Ant Design Blazor, Material Design Blazor and so on.

![image](https://user-images.githubusercontent.com/7299249/127980781-02640db3-791b-4acb-b23e-8ca1d52bb0d2.png)

----
What's the difference with Blazorise?  

The BlazorComponent don't provide any css. You can use Ant Design, Material Design and so on.
It preserves the design philosophy of each design and there is no need to compromise by switch design, but the user switches design at a cost.

----
How to do it?
* High Reusability: 
  * Refactor and abstract the more basic components. 
  * _[Welcome to supplement]_
* High Stability:
  * Improve unit test coverage.
  * Provide global exception handler
  * _[Welcome to supplement]_
* High Performance:
  * More reasonable lazy loading
  * Optimized UI rendering performance
  * Support keep alive
  * Support virtualization, like virtual scroller.
  * _[Welcome to supplement]_
* High Plasticity:
  * Slots everywhere
  * _[Welcome to supplement]_

----
What have we done?

We are based on ```Ant Design Blazor``` and ```Vuetify``` css, developed a MASA Blazor component library, the current completion is only about 60%. We have long term support and a clear long term roadmap.
The repo url: https://github.com/BlazorComponent/MASA.Blazor
Some screenshots(sorry, there is no english document now, but we will support i18n):
![image](https://user-images.githubusercontent.com/7299249/127976948-9f6a62f1-25a8-404a-b248-786f02208f88.png)
![image](https://user-images.githubusercontent.com/7299249/127977078-0ecccfcb-12f7-4138-ac6c-0a823092f90e.png)
![image](https://user-images.githubusercontent.com/7299249/127977150-7afaa8aa-0bef-4b90-881d-7297c73f0bf9.png)
![image](https://user-images.githubusercontent.com/7299249/127977270-ed3dae50-9450-4dfa-abd4-b3b99d5dda84.png)
![image](https://user-images.githubusercontent.com/7299249/127977635-807f1ba2-d8fb-4f14-a172-1e5b3c4814a5.png)

**Comment by ElderJames** [comment]
This is a good idea. It just so happens that Ant Design React has also separated out the [react-component](https://github.com/react-component).

Based on this base component we can use the valuable experience gained by Ant Design Blazor to develop rich component libraries such as Bootstrap 5 or Element.

I have a few questions:
- Firstly, why not implement your needs by way of a custom theme?
- Then, this is a very large amount of work and it is a question of whether more people will come and contribute.
- Finally, will Ant Design Blazor be able to migrate smoothly to this? How much improvement can we get from this?

**Reply by doddgu** [reply]
> Firstly, why not implement your needs by way of a custom theme?

We are using Material Design, it has many templates, we can do this quickly by adjusting the layout.

> Then, this is a very large amount of work and it is a question of whether more people will come and contribute.

We expect to put two people on unit test coverage and two on refactoring, finally leave one for long-term support.

> Finally, will Ant Design Blazor be able to migrate smoothly to this? 

It may not be easy, but we'll write unit tests before migrate.

> How much improvement can we get from this?

High Reusability, High Stability, High Performance, High Plasticity, and even richer base components, such as virtualization.

----
Is there anything else that needs further discussion?

**Comment by theilgaz** [comment]
Great idea. Can you give us a preview of the showcase? 

Please consider that each design system renders ui components differently. Additionally, their style structures are totally different. This should also be considered by the switch mechanism.

**Reply by doddgu** [reply]
We tried to migrate the Icon component and confirmed the refactoring direction after initial communication with James. The code can be found at https://github.com/BlazorComponent/ant-design-blazor/commit/2c15876c564603812eda297f63f285cf8a42cae1

Our repo url is:
https://github.com/BlazorComponent/BlazorComponent
https://github.com/BlazorComponent/MASA.Blazor

Our code is in the develop branch, and the first version is expected to be released at the end of November, and the branch will switch back to main.

You can preview the docs site: https://masa-blazor-docs-dev.lonsid.cn/
This site will show you our `MASA Blazor` component library developed based on `BlazorComponent`.

Finally, we try to make `BlazorComponent` compatible with `Ant Design Blazor`. But you know, some attribute names and parameter values need to be thought again.

**Reply by theilgaz** [reply]
Good work. I think you should add some English readme for existing repo. I'll be back when the first version released.

**Reply by doddgu** [reply]
> Good work. I think you should add some English readme for existing repo. I'll be back when the first version released.

Yeah, we have plans to release the English version.

**Reply by doddgu** [reply]
> Good work. I think you should add some English readme for existing repo. I'll be back when the first version released.

Hi @theilgaz , we have released the first version `0.3.0`, and it has been stable for a while.  And we'll be releasing 0.4.0 soon. Details can be found here https://github.com/BlazorComponent/MASA.Blazor
